File: core/src/main/java/heroes/journey/initializers/base/actions/TravelActions.java
package heroes.journey.initializers.base.actions;

import static heroes.journey.initializers.base.Map.KINGDOM;
import static heroes.journey.initializers.base.Map.TOWN;
import static heroes.journey.initializers.base.actions.BaseActions.popup;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Objects;
import java.util.Queue;
import java.util.UUID;

import heroes.journey.GameState;
import heroes.journey.components.NamedComponent;
import heroes.journey.components.PositionComponent;
import heroes.journey.components.character.ActionComponent;
import heroes.journey.components.character.MapComponent;
import heroes.journey.components.character.MovementComponent;
import heroes.journey.entities.Position;
import heroes.journey.entities.actions.Action;
import heroes.journey.entities.actions.Cost;
import heroes.journey.entities.actions.ShowAction;
import heroes.journey.entities.actions.TargetAction;
import heroes.journey.entities.actions.inputs.ActionInput;
import heroes.journey.entities.actions.inputs.TargetInput;
import heroes.journey.entities.actions.results.ActionListResult;
import heroes.journey.entities.actions.results.ActionResult;
import heroes.journey.entities.actions.results.MultiStepResult;
import heroes.journey.entities.actions.results.StringResult;
import heroes.journey.initializers.InitializerInterface;
import heroes.journey.initializers.utils.Utils;
import heroes.journey.registries.FeatureManager;
import heroes.journey.tilemap.Feature;
import heroes.journey.tilemap.FogUtils;
import heroes.journey.ui.HUD;
import heroes.journey.utils.Direction;
import heroes.journey.utils.ai.pathfinding.Cell;
import heroes.journey.utils.ai.pathfinding.EntityCursorPathing;

public class TravelActions implements InitializerInterface {

    public static Action travel;
    public static TargetAction<UUID> journey, wayfare;
    public static TargetAction<Direction> explore;
    // TODO Pilgrimage lose a turn but go anywhere?
    // TODO No direction explore that expands in a circle?

    // TODO FIX JOURNEY
    @Override
    public void init() {
        explore = new TargetAction<Direction>("explore", "Explore", "Explore in a direction",
            new Cost(5, 2, 0, 0)) {
            @Override
            public List<Direction> getTargets(ActionInput input) {
                return List.of(Direction.getDirections());
            }

            @Override
            public String getTargetDisplayName(TargetInput<Direction> input) {
                return "Explore " + input;
            }

            @Override
            protected ActionResult onSelectTarget(TargetInput<Direction> input) {
                GameState gs = input.getGameState();
                UUID e = input.getEntityId();
                PositionComponent position = PositionComponent.get(gs.getWorld(), e);
                MapComponent mapComponent = MapComponent.get(gs.getWorld(), e);

                // TODO Based on Stamina/Body
                int revealDistance = 8; // How far the cone extends
                // TODO Based on Vision
                int baseWidth = 0;       // Start width (0 = just 1 tile at origin)

                FogUtils.revealCone(gs, mapComponent, position.getX(), position.getY(), revealDistance,
                    baseWidth, input.getInput());

                return new StringResult("You have explored the " + input.getInput());
            }
        }.register();
        // TODO clean up journey to use more utils?
        journey = new TargetAction<UUID>("journey", "Journey", "Travel to a known location",
            new Cost(1, 0, 0, 0)) {
            @Override
            public List<UUID> getTargets(ActionInput input) {
                UUID currentLocation = Utils.getLocation(input);
                MapComponent mapComponent = MapComponent.get(input.getGameState().getWorld(),
                    input.getEntityId());
                List<UUID> journeyLocations = new ArrayList<>();
                for (UUID knownLocation : mapComponent.getKnownLocations()) {
                    if (!Objects.equals(currentLocation, knownLocation))
                        journeyLocations.add(knownLocation);
                }
                return journeyLocations;
            }

            @Override
            public String getTargetDisplayName(TargetInput<UUID> input) {
                return NamedComponent.get(input.getGameState().getWorld(), input.getInput(),
                    "Unknown Location");
            }

            @Override
            protected void onHoverTarget(TargetInput<UUID> input) {
                Feature feature = FeatureManager.getFeature(input.getInput());
                HUD.get()
                    .getCursor()
                    .setMapPointerLoc(new Position(feature.location.getX(), feature.location.getY()));
            }

            @Override
            protected ActionResult onSelectTarget(TargetInput<UUID> input) {
                GameState gs = input.getGameState();
                UUID e = input.getEntityId();
                Feature feature = FeatureManager.getFeature(input.getInput());
                String locationName = NamedComponent.get(gs.getWorld(), input.getInput(), "Unknown Location");
                PositionComponent positionComponent = PositionComponent.get(gs.getWorld(), e);
                Cell path = new EntityCursorPathing().getPath(gs.getMap(), positionComponent.getX(),
                    positionComponent.getY(), feature.location.getX(), feature.location.getY(), e);

                Queue<Runnable> events = new LinkedList<>();
                events.add(() -> {
                    gs.getWorld().edit(e).create(MovementComponent.class).path(path.reverse());
                });
                events.add(() -> {
                    BaseActions.popupMessage = "You have traveled to " + locationName;
                    gs.getWorld().edit(e).create(ActionComponent.class).action(popup);
                });
                return new MultiStepResult(events);
            }

            @Override
            protected ActionResult onSelectAITarget(TargetInput<UUID> input) {
                GameState gs = input.getGameState();
                UUID e = input.getEntityId();
                Feature feature = FeatureManager.getFeature(input.getInput());
                String locationName = NamedComponent.get(gs.getWorld(), input.getInput(), "Unknown Location");
                PositionComponent positionComponent = PositionComponent.get(gs.getWorld(), e);
                Cell path = new EntityCursorPathing().getPath(gs.getMap(), positionComponent.getX(),
                    positionComponent.getY(), feature.location.getX(), feature.location.getY(), e);

                Cell end = gs.getEntities().moveEntity(e, path);
                positionComponent.setPos(end.x, end.y);
                positionComponent.sync();
                return new StringResult("You have traveled to " + locationName);
            }
        }.register();
        wayfare = new TargetAction<UUID>("wayfare", "Wayfare", "Travel to a connected location",
            new Cost(2, 0, 0, 0)) {
            @Override
            public List<UUID> getTargets(ActionInput input) {
                UUID featureId = Utils.getLocation(input);
                Feature feature = FeatureManager.get().get(featureId);
                List<UUID> wayfareLocations = new ArrayList<>();
                for (UUID connectionId : feature.getConnections()) {
                    Feature connection = FeatureManager.getFeature(connectionId);
                    wayfareLocations.add(connection.entityId);
                }
                return wayfareLocations;
            }

            @Override
            public ShowAction internalRequirementsMet(ActionInput input) {
                UUID featureId = Utils.getLocation(input);
                Feature feature = FeatureManager.get().get(featureId);
                return feature.getType() == KINGDOM || feature.getType() == TOWN ?
                    ShowAction.YES :
                    ShowAction.GRAYED;
            }

            @Override
            public String getTargetDisplayName(TargetInput<UUID> input) {
                return NamedComponent.get(input.getGameState().getWorld(), input.getInput(),
                    "Unknown Location");
            }

            @Override
            protected void onHoverTarget(TargetInput<UUID> input) {
                Feature feature = FeatureManager.getFeature(input.getInput());
                HUD.get()
                    .getCursor()
                    .setMapPointerLoc(new Position(feature.location.getX(), feature.location.getY()));
            }

            @Override
            protected ActionResult onSelectTarget(TargetInput<UUID> input) {
                GameState gs = input.getGameState();
                UUID e = input.getEntityId();
                Feature feature = FeatureManager.getFeature(input.getInput());
                String locationName = NamedComponent.get(gs.getWorld(), input.getInput(), "Unknown Location");
                PositionComponent positionComponent = PositionComponent.get(gs.getWorld(), e);
                Cell path = new EntityCursorPathing().getPath(gs.getMap(), positionComponent.getX(),
                    positionComponent.getY(), feature.location.getX(), feature.location.getY(), e);

                Queue<Runnable> events = new LinkedList<>();
                events.add(() -> {
                    gs.getWorld().edit(e).create(MovementComponent.class).path(path.reverse());
                });
                events.add(() -> {
                    BaseActions.popupMessage = "You have traveled to " + locationName;
                    gs.getWorld().edit(e).create(ActionComponent.class).action(popup);
                });
                return new MultiStepResult(events);
            }

            @Override
            protected ActionResult onSelectAITarget(TargetInput<UUID> input) {
                GameState gs = input.getGameState();
                UUID e = input.getEntityId();
                Feature feature = FeatureManager.getFeature(input.getInput());
                String locationName = NamedComponent.get(gs.getWorld(), input.getInput(), "Unknown Location");
                PositionComponent positionComponent = PositionComponent.get(gs.getWorld(), e);
                Cell path = new EntityCursorPathing().getPath(gs.getMap(), positionComponent.getX(),
                    positionComponent.getY(), feature.location.getX(), feature.location.getY(), e);

                Cell end = gs.getEntities().moveEntity(e, path);
                positionComponent.setPos(end.x, end.y);
                positionComponent.sync();
                return new StringResult("You have traveled to " + locationName);
            }
        }.register();

        List<Action> travelActionOptions = new ArrayList<>();
        travelActionOptions.add(explore);
        travelActionOptions.add(wayfare);
        travelActionOptions.add(journey);
        travel = new Action("travel", "Travel", "Choose a travel option", true, null) {
            @Override
            public ActionResult internalOnSelect(ActionInput input) {
                return new ActionListResult(travelActionOptions);
            }
        }.register();
    }
}

File: core/src/main/java/heroes/journey/initializers/base/actions/TravelActions.kt
package heroes.journey.initializers.base.actions

import heroes.journey.components.PositionComponent
import heroes.journey.components.character.ActionComponent
import heroes.journey.components.character.MovementComponent
import heroes.journey.entities.Position
import heroes.journey.entities.actions.Action
import heroes.journey.entities.actions.TargetAction
import heroes.journey.entities.actions.action
import heroes.journey.entities.actions.results.MultiStepResult
import heroes.journey.entities.actions.results.StringResult
import heroes.journey.entities.actions.targetAction
import heroes.journey.initializers.InitializerInterface
import heroes.journey.initializers.utils.Utils
import heroes.journey.registries.Registries.RegionManager
import heroes.journey.tilemap.Region
import heroes.journey.ui.HUD
import heroes.journey.utils.ai.pathfinding.EntityCursorPathing
import java.util.*

class TravelActions : InitializerInterface {
    // TODO Pilgrimage lose a turn but go anywhere?
    override fun init() {
        go = action {
            id = "travel_to"
            name = "Travel to"
            description = "Travel to "
            inputDisplayNameFn = { input ->
                description + RegionManager[input["target"]]!!.getName()
            }
            onHoverFn = { input ->
                val region: Region = RegionManager[input["target"]]!!
                HUD.get()
                    .cursor
                    .setMapPointerLoc(Position(region.center.x, region.center.y))
            }
            onSelectFn = { input ->
                val gs = input.gameState
                val e = input.entityId
                val region: Region = RegionManager[input["target"]]!!
                val positionComponent = PositionComponent.get(gs.world, e)
                val path = EntityCursorPathing().getPath(
                    gs.map, positionComponent.x,
                    positionComponent.y, region.center.x, region.center.y, e
                )

                val events: Queue<Runnable> = LinkedList()
                events.add(Runnable {
                    gs.world.edit(e).create(
                        MovementComponent::class.java
                    ).path(path.reverse())
                })
                events.add(Runnable {
                    val inputs: HashMap<String, String> = HashMap()
                    inputs["message"] = "You have traveled to $input['target']"
                    gs.world.edit(e).create<ActionComponent>(ActionComponent::class.java)
                        .action(BaseActions.popup, inputs)
                })
                MultiStepResult(events)
            }
            onSelectAIFn = { input ->
                val gs = input.gameState
                val e = input.entityId
                val region: Region = RegionManager[input["target"]]!!
                val positionComponent = PositionComponent.get(gs.world, e)
                val path = EntityCursorPathing().getPath(
                    gs.map, positionComponent.x,
                    positionComponent.y, region.center.x, region.center.y, e
                )

                val end = gs.entities.moveEntity(e, path)
                positionComponent.setPos(end.x, end.y)
                positionComponent.sync()
                StringResult("You have traveled to ${region.id}")
            }
        }.register()
        travel = targetAction<Region> {
            id = "travel"
            name = "Travel"
            description = "Travel to a connected location"
            getTargets = { input ->
                val regionId = Utils.getRegion(input)
                val region = RegionManager[regionId]
                val wayfareLocations = ArrayList<Region>()
                for (connectionId in region!!.neighborRegionIds) {
                    val connection = RegionManager[connectionId]!!
                    wayfareLocations.add(connection)
                }
                wayfareLocations
            }
            targetAction = go!!.id
        }.register()
    }

    companion object {
        @JvmField
        var travel: TargetAction<Region>? = null
        var go: Action? = null
    }
}
